civet/compilers/base_compiler.py

```diff
@@ -0,0 +1,217 @@
+from __future__ import print_function
+import atexit
+from distutils.spawn import find_executable
+import os
+import subprocess
+import sys
+
+from django.conf import settings
+from watchdog.events import FileSystemEventHandler
+from watchdog.observers import Observer
+
+from civet.util import collect_src_dst_dir_mappings
+from civet.util import raise_error_or_kill
+
+
+class CompilerObserver(Observer):
+    """Watch source files and compile them on change.
+
+    We have to roll our own watchdog-based solution because:
+
+    1. Unlike sass, coffee does not allow watching multiple directories. This
+       leaves us with only one option: Watch the entire project root
+       directory. That is not viable because we use a different directory
+       layout for the compiled assets (think how collectstatic works).
+    2. coffee can't handle the number of source files we have! This is caused
+       by the combination of node.js's FS watcher implementation and OS X's
+       default limit on the number of open files. This can be mitigated by
+       asking all our devs to remember to dial up the limit manually, but then
+       again 1. makes it hard to work with. For details, see
+       https://github.com/joyent/node/issues/2479
+    """
+
+    def start(self):
+        super(CompilerObserver, self).start()
+
+        # Stop the observer when Django's autoreload calls sys.exit() before
+        # reloading
+        def cleanup():
+            self.stop()
+
+        atexit.register(cleanup)
+
+
+class CompilerFSEventHandler(FileSystemEventHandler):
+    """A watchdog FS event handler for watching file changes and compiling.
+
+    We don't handle directory creation events anywhere in our source dir. We
+    rarely add new directories, and when that happens, we can always re-start
+    the runserver command. Handling directory events correctly will require
+    more code than is practical.
+    """
+
+    def __init__(self, compiler, src_dst_dir_map):
+        self.compiler = compiler
+        super(FileSystemEventHandler, self).__init__()
+        self.src_dst_dir_map = src_dst_dir_map
+
+    def get_dst_path(self, src_path):
+        src_dir, src_filename = os.path.split(src_path)
+        dst_dir = self.src_dst_dir_map.get(src_dir)
+        if not dst_dir:
+            return None
+        dst_filename = os.path.splitext(src_filename)[0] + '.js'
```

Comment:
This doesn't seem right, since you're always appending the `.js` extension. It seems like this isn't causing a bug now because `SassCompiler` overrides the `watch()` method and doesn't `CompilerFSEventHandler`, but we should fix this because it's a bit misleading.